gremlin-dotnet/src/Gremlin.Net/Driver/ConnectionPool.cs

```diff
@@ -66,42 +61,35 @@ public async Task<IConnection> GetAvailableConnectionAsync()
 
         private async Task EnsurePoolIsPopulatedAsync()
         {
-            // The pool could have been empty because of connection problems. So, we need to populate it again.
+            // The pool could have been (partially) empty because of connection problems. So, we need to populate it again.
```

Comment:
I think we should implement some kind of back-off or delayed retry. Otherwise, if there is a failure, this loop will never stop until we get it to the pool gets to the final length.

Reply:
So, a counter that breaks out of the loop after like 10 retries and `Task.Delay(TimeSpan.FromMilliseconds(5))` before the `continue`?

Reply:
Depends on how we want to handle pool failures.

From what I understand, `GetAvailableConnectionAsync()` gets called every time, and it calls `EnsurePoolIsPopulatedAsync()` so, it's possible that during the lifetime of the driver, a connection is closed and `EnsurePoolIsPopulatedAsync()` needs to create a new one:
- If there is another connection open, I would prefer `EnsurePoolIsPopulatedAsync()` to occur in the background, with some kind of back-off delay (in the order of hundreds of millis, seconds or we can use exponential back off).
- If there isn't another open connection, we can await until one connection is open and fail as fast as possible.

Reply:
> If there is another connection open, I would prefer EnsurePoolIsPopulatedAsync() to occur in the background

That is what you suggested with the task scheduler, right? (Maybe I'm just understanding you wrong here, I'm especially not sure about how a back-off delay could work.) I agree that doing this in the background makes more sense in this case, but I won't be able to implement this in time for 3.4.1 and I also prefer not to change too much at the same time if it's not necessary.
So, I would like to just create a ticket for this issue and then try to implement it for 3.4.2.

Adding a delay and a counter here to the loop would be easy to implement. So, if you think that's already an improvement then I can add that.

> fail as fast as possible.

That should be the current situation as we don't catch any exception here. So it fails when one connection establishment fails.